Generals/Code/GameEngine/Source/GameClient/InGameUI.cpp

```diff
@@ -5424,42 +5424,37 @@ void InGameUI::selectNextIdleWorker( void )
 	if(getSelectCount() == 0 || getSelectCount() > 1)
 	{
 		selectThisObject = *m_idleWorkers[index].begin();
+		// If our idle worker is contained by anything, we need to select the container instead.
+		while (selectThisObject->getContainedBy())
+			selectThisObject = selectThisObject->getContainedBy();
 	}
 	else
 	{
 		Drawable *selectedDrawable = TheInGameUI->getFirstSelectedDrawable();	
-		
-		ObjectListIt it = m_idleWorkers[index].begin();
-		while(it != m_idleWorkers[index].end())
+		std::vector<Object*> uniqueIdleWorkers = getUniqueIdleWorkers(m_idleWorkers[index]);
+
+		ObjectPtrVector::iterator uit = uniqueIdleWorkers.begin();
+		while(uit != uniqueIdleWorkers.end())
 		{
-			Object *itObj = *it;
+			Object *itObj = *uit;
 			if(itObj == selectedDrawable->getObject())
 			{
-				++it;
-				if(it != m_idleWorkers[index].end())
-					selectThisObject = *it;
+				++uit;
+				if(uit != uniqueIdleWorkers.end())
+					selectThisObject = *uit;
 				else
-					selectThisObject = *m_idleWorkers[index].begin();
+					selectThisObject = *uniqueIdleWorkers.begin();
 				break;
 			}
-			++it;
+			++uit;
 		}
 		// if we had something selected that wasn't a worker, we'll get here
-		if(!selectThisObject)
-			selectThisObject = *m_idleWorkers[index].begin();
-
+		if (!selectThisObject)
+			selectThisObject = uniqueIdleWorkers.front();
 	}
 	DEBUG_ASSERTCRASH(selectThisObject, ("InGameUI::selectNextIdleWorker Could not select the next IDLE worker"));
 	if(selectThisObject)
 	{	
-		
-		//If our idle worker is contained by anything, we need to select the container instead.
-		Object *containedBy = selectThisObject->getContainedBy();
```

Comment:
Can add here `DEBUG_ASSERTCRASH(selectThisObject->getContainedBy() == NULL, ("Selected object is not expected to be contained"));`